js/base-engine.js

```diff
@@ -203,6 +204,70 @@ function d20plusEngine () {
 		});
 	};
 
+	d20plus.engine.enhanceMacros = () => {
+		const $dialog = $(`.dialog[data-macroid=${d20plus.engine._lastOpenedMacroId}]`);
+		if (!d20plus.engine._lastOpenedMacroId || !$dialog[0]) return;
+		const $macro = $dialog.find(`.macro.tokenizer`);
+		const $name = $dialog.find("input.name");
+		const $checkbox = $dialog.find(".isjs")
+			.on("change", (evt) => {
+				if ($(evt.target).prop("checked")) $macro.parent().addClass("jsdialog");
+				else $macro.parent().removeClass("jsdialog");
+			});
+		const script = d20plus.engine.decodeScript($macro.val());
+		if (script) {
+			$macro.parent().addClass("jsdialog");
+			$macro.val(script);
+			$checkbox.prop("checked", true);
+		}
+		$dialog.find(".btn.testmacro").on("click", () => {
+			if (!$checkbox.prop("checked")) return;
+			const script = $macro.val();
+			$macro.val(d20plus.engine.runScript(script));
+			setTimeout(() => $macro.val(script), 100);
+		});
+		const $buttons = $dialog.parent()
+			.find(".ui-dialog-buttonpane button:not(.active)")
+			.addClass("active");
+		$buttons.on("mouseup", (evt) => {
+			if (!$name.val()) {
+				let i = 0;
+				while (d20plus.ut.getMacroByName(`Untitled${i}`)) i++;
+				$name.val(`Untitled${i}`);
+			}
+			if (!$checkbox.prop("checked") || evt.which !== 1) return;
+			$macro.val(d20plus.engine.encodeScript($macro.val()));
+		});
+	}
+
+	d20plus.engine.decodeScript = (macro) => {
+		const parts = macro.split("...");
+		if (parts.length !== 3
+			|| parts[0] !== "bs``<``"
```

Comment:
q: what does "bs" stand for, "betteR20 script"?

Reply:
that's one of the possible interpretations, yes 😄